integration/kotlinx-coroutines-jdk8/src/future/Future.kt

```diff
@@ -137,34 +138,98 @@ public fun <T> CompletionStage<T>.asDeferred(): Deferred<T> {
 }
 
 /**
- * Awaits for completion of the completion stage without blocking a thread.
+ * Converts this [CompletableFuture] to an instance of [Deferred].
+ * The future is cancelled when the resulting deferred is cancelled.
+ *
+ * ### Implementation details
+ *
+ * This implementation has a fast-path for a case of a future that [isDone][CompletableFuture.isDone].
+ */
+@Suppress("DeferredIsResult")
+public fun <T> CompletableFuture<T>.asDeferred(): Deferred<T> {
+    // Fast path if already completed
+    if (isDone) {
+        return try {
+            @Suppress("UNCHECKED_CAST")
+            CompletableDeferred(get() as T)
+        } catch (e: Throwable) {
+            // unwrap original cause from ExecutionException
+            val original = (e as? ExecutionException)?.cause ?: e
+            CompletableDeferred<T>().also { it.completeExceptionally(original) }
+        }
+    }
+    // slow-path
+    return (this as CompletionStage<T>).asDeferred()
+}
+/**
+ * Awaits for completion of [CompletionStage] without blocking a thread.
  *
  * This suspending function is cancellable.
  * If the [Job] of the current coroutine is cancelled or completed while this suspending function is waiting, this function
  * stops waiting for the completion stage and immediately resumes with [CancellationException][kotlinx.coroutines.CancellationException].
- * This method is intended to be used with one-shot futures, so on coroutine cancellation completion stage is cancelled as well if it is instance of [CompletableFuture].
+ * This method is intended to be used with one-shot futures, so on coroutine cancellation completion stage is cancelled as well if it is instance of a [Future].
  * If cancelling given stage is undesired, `stage.asDeferred().await()` should be used instead.
+ *
+ * ### Implementation details
```

Comment:
Actually, there is far more elegant and less-error prone solution: `CompletableStage.toCompletableFuture()`.
It returns `this` for actual instances of CF and does the corresponding wrapping for `MinimalStage` including proper `cancel` propagation.

This also will allow us to leave the single method and users can safely upcast the type without `UOE`  from our side

Reply:
Very good 👍